### PR TITLE
Support coprocessor

### PR DESCRIPTION
**Pull request by liukun4515**
Implement the insert and read operation in the coprocessor.

### CODE REVIEW

File: db/tikv/coprocessor.go [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+	return make(map[string][]byte, len(fields)), nil
+}
+
+func (db *coprocessor) Scan(ctx context.Context, table string, startKey string, count int, fields []string) ([]map[string][]byte, error) {
+
+	return nil, errors.New("unsupported type of scan in coprocessor")
+}
+
+func (db *coprocessor) Update(ctx context.Context, table string, key string, values map[string][]byte) error {
+	return errors.New("unsupported type of update in coprocessor")
```
Comment (siddontang) [inline db/tikv/coprocessor.go]: seem you can use txn API to support update and delete, why not support them?

File: pkg/util/table.go [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+	output := make([]uint32, 0, len(fields))
+	for i := 0; i < len(fields); i++ {
+		output = append(output, uint32(i))
+	}
+	dag.OutputOffsets = output
+	dag.Executors = []*tipb.Executor{t.getTableScanExe(output)}
+	return dag
+}
+
+func (t *Table) getTableScanExe(indexes []uint32) *tipb.Executor {
```
Comment (siddontang) [inline pkg/util/table.go]: getTableScanExec

File: pkg/util/table.go [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+			Tp:        int32(mysql.TypeVarchar)})
+	}
+	return table
+}
+
+func (t *Table) DAGTableScanReq(fields []string) *tipb.DAGRequest {
+	dag := &tipb.DAGRequest{}
+	dag.StartTs = math.MaxInt64
+	output := make([]uint32, 0, len(fields))
+	for i := 0; i < len(fields); i++ {
```
Comment (siddontang) [inline pkg/util/table.go]: For YCSB, fields can be empty which means fetching all the fields.

File: pkg/util/table.go [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+
+package util
+
+import (
+	"github.com/magiconair/properties"
+	"github.com/pingcap/tidb/kv"
+	"github.com/pingcap/tidb/mysql"
+	"github.com/pingcap/tidb/tablecodec"
+	"github.com/pingcap/tidb/types"
+	"github.com/pingcap/tipb/go-tipb"
```
Comment (siddontang) [inline pkg/util/table.go]: please use the vendor to save the dependence.

**Comment by siddontang** [comment]
PTAL @breeswish

File: db/tikv/coprocessor.go (hunk 1 above)
Comment (liukun4515) [reply db/tikv/coprocessor.go]: I think the coprocessor just need to support the query, so I didn't support update and delete.
I will add update and delete later if needed.

File: db/tikv/coprocessor.go [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+	"github.com/magiconair/properties"
+	"github.com/pingcap/go-ycsb/pkg/util"
+	"github.com/pingcap/go-ycsb/pkg/ycsb"
+	"github.com/pingcap/tidb/kv"
+	"github.com/pingcap/tidb/store/tikv"
+	"github.com/pkg/errors"
+)
+
+func createCoprocessorDB(p *properties.Properties) (ycsb.DB, error) {
+	pdAddr := p.GetString(tikvPD, "127.0.0.1:2379")
```
[the remark this answers is not in the document]
Comment (liukun4515) [reply db/tikv/coprocessor.go]: The pd address is configurable, you can add `-p tikv.pd=xxxxxx:xxxx` to set the address. The `127.0.0.1:2379` is the default value if you don't give your pd address.

File: pkg/util/table.go (hunk 3 above)
Comment (liukun4515) [reply pkg/util/table.go]: Hava supported the empty fields

File: pkg/util/table.go (hunk 4 above)
Comment (liukun4515) [reply pkg/util/table.go]: all the dependencies were saved by using the vendor.

File: pkg/util/table.go [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+			ColumnId:  int64(i),
+			Collation: int32(mysql.DefaultCollationID),
+			ColumnLen: types.UnspecifiedLength,
+			Decimal:   types.UnspecifiedLength,
+			Tp:        int32(mysql.TypeVarchar)})
+	}
+	return table
+}
+
+func (t *Table) DAGTableScanReq(fields []string) *tipb.DAGRequest {
```
[the remark this answers is not in the document]
Comment (liukun4515) [reply pkg/util/table.go]: ok

File: cmd/go-ycsb/main.go [hunk 7]
```diff
@@ -99,7 +99,7 @@ func initialGlobal(dbName string, onProperties func()) {
 	if globalDB, err = dbCreator.Create(globalProps); err != nil {
 		util.Fatalf("create db %s failed %v", dbName, err)
 	}
-
+	fmt.Println(globalProps.String())
```
[the remark this answers is not in the document]
Comment (liukun4515) [reply cmd/go-ycsb/main.go]: ok

File: pkg/util/table.go [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+	for i := 0; i < len(fields); i++ {
+		output = append(output, uint32(i))
+	}
+	dag.OutputOffsets = output
+	dag.Executors = []*tipb.Executor{t.getTableScanExe(output)}
+	return dag
+}
+
+func (t *Table) getTableScanExe(indexes []uint32) *tipb.Executor {
+	exe := &tipb.Executor{}
```
[the remark this answers is not in the document]
Comment (liukun4515) [reply pkg/util/table.go]: ok

File: db/tikv/coprocessor.go [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+	if err != nil {
+		return nil, err
+	}
+	//return db.table.DecodeValue(result.GetData(), fields)
+	return make(map[string][]byte, len(fields)), nil
+}
+
+func (db *coprocessor) Scan(ctx context.Context, table string, startKey string, count int, fields []string) ([]map[string][]byte, error) {
+
+	return nil, errors.New("unsupported type of scan in coprocessor")
```
[the remark this answers is not in the document]
Comment (liukun4515) [reply db/tikv/coprocessor.go]: If we want to use the scan in coprocessor, I should construct the complexed executor (scan + limit). But I'm not similar to these, I need more time to read the code in TiDB.
I will support the scan and decode the respond of coprocessor in the next pr.

[the remark this answers is not in the document]
Comment (breeswish) [reply db/tikv/coprocessor.go]: That's ok!

**Comment by breeswish** [comment]
LGTM

**Pull request merged**